README.md
# WebClip Element
Take a screenshot of highligthed element.

![screenshot](./webstore_assets/screenshot_01.png "Screenshot")

You can find it here: https://chrome.google.com/webstore/detail/webclipelement/ndkjiolmefobbpbiejfnihoccngfbpjh

# Overview
Click on the extension icon, select an element on the page and click on it.

Voilà... A screenshot of that element will open in a new tab.
Hold SHIFT to clip multiple elements on the page, or press ESC to quit clipping mode.

You can also take a screenshot of the visible page:
right-click on the extension button and select 'Take page screenshot...'.

HOW DOES IT WORK?
Under-the-hood it takes a screenshot of the current (visible) page, and then crops it using position and size of the highlighted element.
Should also work when the browser page is zoomed in/out.

And all happens inside your browser, nothing is sent online (check the code at https://github.com/azrafe7/webClipElement). ;D

CHANGELOG: https://github.com/azrafe7/webClipElement/blob/main/CHANGELOG.md


CHANGELOG.md
 # CHANGELOG

 ### v0.2.4 (22 Oct 2023)
 - update picker to v0.2.4
 - adjust clipping rect
 - support more pages
 
 ### v0.2.0 (15 Oct 2023)
 - update picker to latest
 - inject in top frame only
 - use an iframe overlay when enabled
 - works on sites with multiple frames
 - better close detection
 - better highlight and info

 ### v0.1.6 (25 Sep 2023)
 - disable picker on right click

 ### v0.1.5 (24 Sep 2023)
 - less noise in console (only log if DEBUG == true)
 - change picker cursor when holding SHIFT
 - trigger on "mouseup"

 ### v0.1.3 (21 Sep 2023)
 - add 'take page screenshot...' to action button context menu

 ### v0.1.2 (18 Aug 2023)
 - hold SHIFT to continue clipping
 - set hoverBoxInfo id

 ### v0.1.1 (17 Aug 2023)
 - trigger action on click instead of mouseup
 - update picker
 - add/remove event listeners on enable/disable
 - more reliably hide overlay before taking screenshot

 ### v0.1.0 (13 Aug 2023)
 - enable/disable picker instead of recreating it
 - remove unnecessary permissions
 - make it work on zoomed pages
 - correctly crop only visible part of the element
 - add hoverBoxInfo
 - change overlay color to light green
 - fix bug with elementPicker returning early even if triggered

background.js
"use strict";

let manifest = chrome.runtime.getManifest();
console.log(manifest.name + " v" + manifest.version);

// add contextMenu entry to action button
function createContextMenu() {
  chrome.contextMenus.removeAll(function() {
    chrome.contextMenus.create({
      id: "WebClipElement_onPageScreenshotContextMenu",
      title: "Take page screenshot...",
      contexts: ["action"],
    });
  });
}

chrome.runtime.onInstalled.addListener(() => {
  createContextMenu();
});

// enable picker when clicking the browser action
chrome.action.onClicked.addListener(async (tab) => {
  console.log("[WebClipElement:BG] enablePicker");
  chrome.tabs.sendMessage(
    tab.id,
    {
      event: "enablePicker",
      data: null,
    }
  );
});

async function takeScreenshot() {
  let dataURL = await chrome.tabs.captureVisibleTab(null, {
      format: 'png'
    },
  );
  return dataURL;
}

chrome.runtime.onMessage.addListener(async (msg, sender, sendResponse) => {
  console.log("[WebClipElement:BG]", msg);
  const { event, data } = msg;

  if (event === "takeScreenshot") {
    console.log("[WebClipElement:BG] taking screenshot...");
    let dataURL = await takeScreenshot();
    console.log("[WebClipElement:BG]", dataURL);
    let [activeTab] = await chrome.tabs.query({active: true, lastFocusedWindow: true});
    chrome.tabs.sendMessage(activeTab.id, {event: "takenScreenshot", data:{dataURL: dataURL, hoverInfo: data.hoverInfo, continuePicking: data?.continuePicking}});
  } else if (event === "openCroppedInNewTab") {
    console.log("[WebClipElement:BG] opening cropped image in new tab...");
    let dataURL = data.dataURL;
    let focusNewTab = !(data?.continuePicking);
    let [activeTab] = await chrome.tabs.query({active: true, lastFocusedWindow: true});
    chrome.tabs.create({url: dataURL, index: activeTab.index + 1, active: focusNewTab});
  }
});

chrome.contextMenus.onClicked.addListener(async (info, tab) => {
  console.log("[WebClipElement:BG] onContextMenuClicked:", [info, tab]);

  if (info.menuItemId === "WebClipElement_onPageScreenshotContextMenu") {
    console.log("[WebClipElement:BG] taking page screenshot...");
    chrome.tabs.sendMessage(
      tab.id,
      {
        event: "enablePicker",
        data: { enable: false },
      }
    );
    let dataURL = await takeScreenshot();
    let focusNewTab = true;
    let [activeTab] = await chrome.tabs.query({active: true, lastFocusedWindow: true});
    chrome.tabs.create({url: dataURL, index: activeTab.index + 1, active: focusNewTab});
  }
});


contentScript.js
"use strict";

(async () => {
  
  const DEBUG = false;
  let debug = {
    log: DEBUG ? console.log.bind(console) : () => {} // log or NO_OP
  }

  let manifest = chrome.runtime.getManifest();
  console.log(manifest.name + " v" + manifest.version);

  const HIGHLIGHT_RED = "rgba(250, 70, 60, 0.5)";
  const HIGHLIGHT_GREEN = "rgba(17, 193, 12, 0.5)";
  const HIGHLIGHT_BG_COLOR = HIGHLIGHT_GREEN;

  const OUTLINE_RED = "rgba(250, 70, 60, 0.75)";
  const OUTLINE_GREEN = "rgba(17, 193, 12, 0.90)";
  const OUTLINE_COLOR = OUTLINE_GREEN;

  const CURSORS = ["crosshair", "copy"];

  /* if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
    // dark mode
    HIGHLIGHT_BG_COLOR = HIGHLIGHT_DARK;
    OUTLINE_COLOR = OUTLINE_DARK;
  } */

  let options = {
    container: null,
    iFrameId: 'WebClip Element Picker Frame',
    enabled: false,
    selectors: "*",
    background: HIGHLIGHT_BG_COLOR,
    borderWidth: 0,
    outlineWidth: 1,
    outlineColor: OUTLINE_COLOR,
    transition: "",
    ignoreElements: [],
    action: {},
    hoverBoxInfoId: 'webclip_picker_info',
  }

  // create "disabled" elementPicker on page load
  let elementPicker = new ElementPicker(options);

  // elementPicker.hoverBox.style.cursor = CURSORS[0];
  elementPicker.action = {
    trigger: "mouseup",
    
    callback: ((event, target) => {
      // debug.log("[WebClipElement:CTX] event:", event);
      let continuePicking = event.shiftKey;
      if (event.button == 0) { // only proceed if left mouse button was pressed
        debug.log("[WebClipElement:CTX] target:", target);
        debug.log("[WebClipElement:CTX] info:", elementPicker.hoverInfo);
        window.focus();
        elementPicker.hoverInfo.element = null; // not serializable
        const hoverInfoClone = structuredClone(elementPicker.hoverInfo);
        setTimeout(() => { // to ensure picker overlay is removed
          chrome.runtime.sendMessage(
            {
              event: "takeScreenshot",
              data: {hoverInfo: hoverInfoClone, continuePicking: continuePicking},
            },
          );
        }, 50);
      }
      
      elementPicker.enabled = false; // always disable picker highlight (so that it's not saved in the screenshot)
    })
  }

  function getVisibleRect(rect) {
    let visibleRect = DOMRect.fromRect(rect);

    if (visibleRect.x < 0) {
      visibleRect.width += visibleRect.x;
      visibleRect.x = 0;
    }
    if (visibleRect.y < 0) {
      visibleRect.height += visibleRect.y;
      visibleRect.y = 0;
    }
    if (visibleRect.x + visibleRect.width > window.innerWidth) {
      visibleRect.width = window.innerWidth - visibleRect.x;
    }
    if (visibleRect.y + visibleRect.height > window.innerHeight) {
      visibleRect.height = window.innerHeight - visibleRect.y;
    }
    return visibleRect;
  }

  chrome.runtime.onMessage.addListener((msg, sender, sendResponse) => {
    debug.log("[WebClipElement:CTX]", msg);
    const { event, data } = msg;

    if (event === "enablePicker") {
      elementPicker.enabled = data?.enable ?? true;
      elementPicker.hoverBox.style.cursor = CURSORS[0];
    } else if (event === "takenScreenshot") {
      let dataURL = data.dataURL;
      let hoverInfo = data.hoverInfo;
      let continuePicking = data?.continuePicking;
      
      if (continuePicking) {
        elementPicker.enabled = true;
      }
      
      let image = new Image();
      image.onload = () => {
        let rect = {x: hoverInfo.left, y: hoverInfo.top, width: hoverInfo.width, height: hoverInfo.height};
        let visibleRect = getVisibleRect(rect); //getVisibleRect(hoverInfo.clientRect);
        debug.log("[WebClipElement:CTX] cropping...", visibleRect);
        let canvas = document.createElement('canvas');
        let ctx = canvas.getContext('2d');

        const zoomLevel = window.devicePixelRatio;
        if (zoomLevel != 1.0) {
          visibleRect.x *= zoomLevel;
          visibleRect.y *= zoomLevel;
          visibleRect.width *= zoomLevel;
          visibleRect.height *= zoomLevel;
        }
        canvas.width = visibleRect.width;
        canvas.height = visibleRect.height;
        
        ctx.drawImage(image, visibleRect.x, visibleRect.y, visibleRect.width, visibleRect.height,
                             0, 0, visibleRect.width, visibleRect.height);
        
        ((croppedDataURL) => {
          canvas = null;
          ctx = null;
          debug.log("[WebClipElement:CTX] send cropped dataURL", croppedDataURL);
          chrome.runtime.sendMessage(
            {
              event: "openCroppedInNewTab",
              data: {
                dataURL: croppedDataURL,
                continuePicking: continuePicking,
              },
            },
          );
        })(canvas.toDataURL());
      };
      image.src = dataURL;
    }
  });

  // close picker when pressing ESC
  window.addEventListener('keyup', function(e) {
    if (e.keyCode == 27 && elementPicker.enabled) {
      elementPicker.enabled = false;
      debug.log("[WebClipElement:CTX] user aborted");
    }
  }, true);

  // change picker cursor when holding SHIFT
  function updateCursor(eventInfo) {
    let {keyUp, event} = eventInfo;
    if (elementPicker.enabled) {
      let cursorIdx = +event.shiftKey;
      if (elementPicker.hoverBox.style.cursor != CURSORS[cursorIdx]) {
        // debug.log('[WebClipElement:CTX] change cursor to ' + CURSORS[cursorIdx]);
        elementPicker.hoverBox.style.cursor = CURSORS[cursorIdx];
      }
    }
  }
  window.addEventListener('keyup', (e) => updateCursor({keyUp: true, event: e}), true);
  window.addEventListener('keydown', (e) => updateCursor({keyUp: false, event: e}), true);

})();


ElementPicker.js
(function () {
    function ellipsize(text, length) {
      if (text.length > length) {
        return `${text.substring(0, length)}...${text.substring(text.length - length)}`; 
      } else {
        return text;
      }
    }

    class ElementPicker {
        VERSION = "0.2.4";
        
        constructor(options) {
            // MUST create hover box first before applying options
            this.hoverBox = document.createElement("div");
            this.hoverBox.style.position = "absolute";
            // this.hoverBox.style.pointerEvents = "none";
            this.hoverBox.style.cursor = "crosshair";
            this.hoverBox.style.setProperty("z-index", 2147483647, "important");
          
            this._actionEvent = null;

            this.hoverBoxInfo = document.createElement("div");
            this.hoverInfo = {
              element: null,
              tagName: "",
              width: 0,
              height: 0,
            }
            this.hoverBoxInfo.innerText = "";
            this.hoverBoxInfo.style = 
              `background-color: rgba(0,0,0,.5);
              border-radius: 0 0 0 0;
              bottom: 0;
              box-shadow: 0 0 1px 0 rgba(0,0,0,.16);
              box-sizing: border-box;
              color: #f1f3f4;
              font-family: Roboto-Medium,Roboto,arial,sans-serif;
              font-size: 10px;
              line-height: 10px;
              margin-left: 0;
              overflow: hidden;
              padding: 4px;
              padding-bottom: 2px;
              position: fixed;
              right: 0;
              white-space: nowrap;
              z-index: 2147483647 !important;
              pointer-events: none;`;
    
            const defaultOptions = {
                container: null, // if falsey an iframe will be used
                iFrameId: null, // only used if container is falsey to name the built iframe
                enabled: true,
                selectors: "*", // default to pick all elements
                background: "rgba(153, 235, 255, 0.5)", // transparent light blue
                borderWidth: 5,
                outlineColor: "rgba(153, 235, 255, 0.75)", // transparent light blue
                outlineWidth: 1,
                transition: "all 150ms ease", // set to "" (empty string) to disable
                ignoreElements: [document.body],
                action: {},
                hoverBoxInfoId: 'EP_hoverBoxInfo',
            }
            const mergedOptions = {
                ...defaultOptions,
                ...options
            };

            if (!mergedOptions.container) {
              let pickerIFrame = document.createElement('iframe');
              pickerIFrame.id = mergedOptions.iFrameId ?? 'picker_iframe';
              document.documentElement.append(pickerIFrame);

              const pickerIFrameCSS = `
                backgroundColor: transparent;
                left: 0px;
                top: 0px;
                position: fixed;
                width: 100% !important;
                height: 100% !important;
                overflow: hidden;
                z-index: 2147483647 !important;
                margin: 0px;
                border: 0px;
                color-scheme: none;
              `;
              
              pickerIFrame.style = pickerIFrameCSS;
              pickerIFrame.contentDocument.body.style = pickerIFrameCSS;
              this.iframe = pickerIFrame;
              mergedOptions.container = pickerIFrame.contentDocument.body;
              mergedOptions.ignoreElements.push(pickerIFrame);
            }

            Object.keys(mergedOptions).forEach((key) => {
                this[key] = mergedOptions[key];
            });

            this._onScroll = (e) => {
              let fakeTarget = document.elementsFromPoint(this._lastClientX, this._lastClientY)[1];
              let fakeEvent = {
                clientX: this._lastClientX,
                clientY: this._lastClientY,
                target: fakeTarget,
              }
              this._detectMouseMove(fakeEvent);
            }
            
            this._detectMouseMove = (e) => {
                this._lastClientX = e.clientX;
                this._lastClientY = e.clientY;
                if (!this.enabled) return;
                this._previousEvent = e;
                let target = e.target;
                // console.log("TCL: ElementPicker -> this._moveHoverBox -> target", target)
                if (this.ignoreElements.indexOf(target) === -1 && target.matches(this.selectors)) { // is NOT in ignored elements
                    // console.log("TCL: target", target);
                    if (target === this.hoverBox || target === this.container) {
                        // the truely hovered element behind the added hover box
                        const hoveredElements = document.elementsFromPoint(e.clientX, e.clientY);
                        // console.log(hoveredElements);
                        let hoveredElement = hoveredElements[0];
                        for (hoveredElement of hoveredElements) {
                          if ((this.iframe && this.iframe.contains(hoveredElement)) || this.container.contains(hoveredElement)) {
                            continue;
                          } else {
                            break;
                          }
                        }
                        // console.log("screenX: " + e.screenX);
                        // console.log("screenY: " + e.screenY);
                        // console.log("TCL: hoveredElement", hoveredElement);
                        /*if (!this._triggered && this._previousTarget === hoveredElement) {
                            // avoid repeated calculation and rendering
                            return;
                        } else*/ {
                            target = hoveredElement;
                        }
                    } else {
                        this._previousTarget = target;
                    }
                    const targetOffset = target.getBoundingClientRect();
                    const targetHeight = targetOffset.height;
                    const targetWidth = targetOffset.width;

                    this.hoverBox.style.width = targetWidth + this.borderWidth * 2 + "px";
                    this.hoverBox.style.height = targetHeight + this.borderWidth * 2 + "px";
                    
                    this.hoverBox.style.outline = this.outlineWidth + "px solid " + this.outlineColor;
                    
                    // need scrollX and scrollY to account for scrolling
                    const top = (target.tagName === 'HTML' ? 0 : targetOffset.top) + (this.iframe ? 0 : window.scrollY);
                    const left = (target.tagName === 'HTML' ? 0 : targetOffset.left) + (this.iframe ? 0 : window.scrollX);
                    this.hoverBox.style.top = top - this.borderWidth + "px";
                    this.hoverBox.style.left = left - this.borderWidth + "px";

                    // const infoText = `${targetText} ${targetWidth} × ${targetHeight}`;
                    const attrs = Array.from(target.attributes, ({name, value}) => (name + '=' + value));
                    const ellipsizedAttrsText = attrs.length > 0 ? ' ' + ellipsize(attrs.join(' '), 20) : '';
                    const infoText = `<${target.tagName.toUpperCase()}${ellipsizedAttrsText}> ${targetWidth} × ${targetHeight}`;
                    this.hoverBoxInfo.innerText = infoText;

                    this.hoverInfo = {
                      element: target,
                      tagName: target.tagName.toUpperCase(),
                      width: targetWidth,
                      height: targetHeight,
                      targetOffsetTop: targetOffset.top,
                      targetOffsetLeft: targetOffset.left,
                      scrollX: window.scrollX,
                      scrollY: window.scrollY,
                      top: top, // targetOffset.top + window.scrollY,
                      left: left, // targetOffset.left + window.scrollX,
                      clientRect: targetOffset,
                      text: infoText,
                    }

                    // console.log(this.hoverInfo);
                    
                    if (this._triggered && this.action.callback) {
                        // console.log("TRIGGERED");
                        this.action.callback(this._actionEvent, target);
                        this._triggered = false;
                        this._actionEvent = null;
                    }
                } else {
                    // console.log("hiding hover box...");
                    this.hoverBox.style.width = 0;
                }
            };
            
        }
        get info() {
            return this.hoverInfo;
        }
        get hoverBoxInfoId() {
            return this.hoverBoxInfo.id;
        }
        set hoverBoxInfoId(value) {
            this.hoverBoxInfo.id = value;
        }
        get enabled() {
            return this._enabled;
        }
        set enabled(value) {
            this._enabled = value;
            
            this.hoverBox.style.visibility = this._enabled ? "visible" : "hidden";
            this.hoverBoxInfo.style.visibility = this._enabled ? "visible" : "hidden";
            this.container.style.visibility = this._enabled ? "visible" : "hidden";
            if (this.iframe) {
              this.iframe.style.visibility = this._enabled ? "visible" : "hidden";
              this.iframe.style.display = this._enabled ? "block" : "none";
            }
            this._triggered = false;
            // console.log("set enabled:", this._enabled);
            if (!this._enabled) {
              this.hoverBox.style.width = 0;
              this.hoverBox.style.height = 0;
              this.hoverBoxInfo.innerText = '';
              if (this._triggerListener) {
                this.container.removeEventListener(this.action.trigger, this._triggerListener);
              }
              this.container.removeEventListener("mousemove", this._detectMouseMove);
              document.removeEventListener("scroll", this._onScroll);
              // console.log("remove listeners");
            } else {
              window.focus(); // ensure window is focused so it can listen to key events
              if (this.action?.trigger && this._triggerListener) {
                this.container.addEventListener(this.action.trigger, this._triggerListener);
              }
              this.container.addEventListener("mousemove", this._detectMouseMove);
              document.addEventListener("scroll", this._onScroll);
              // console.log("add listeners");
            }
        }
        get container() {
            return this._container;
        }
        set container(value) {
            if (value.appendChild) {
                this._container = value;
                this.container.appendChild(this.hoverBox);
                this.container.appendChild(this.hoverBoxInfo);
            } else {
                throw new Error("Please specify an HTMLElement as container!");
            }
        }
        get background() {
            return this._background;
        }
        set background(value) {
            this._background = value;

            this.hoverBox.style.background = this.background;
        }
        get outlineWidth() {
            return this._outlineWidth;
        }
        set outlineWidth(value) {
            this._outlineWidth = value;

            this._redetectMouseMove();
        }
        get transition() {
            return this._transition;
        }
        set transition(value) {
            this._transition = value;

            this.hoverBox.style.transition = this.transition;
        }
        get borderWidth() {
            return this._borderWidth;
        }
        set borderWidth(value) {
            this._borderWidth = value;

            this._redetectMouseMove();
        }
        get selectors() {
            return this._selectors;
        }
        set selectors(value) {
            this._selectors = value;

            this._redetectMouseMove();
        }
        get ignoreElements() {
            return this._ignoreElements;
        }
        set ignoreElements(value) {
            this._ignoreElements = value;

            this._redetectMouseMove();
        }
        get action() {
            return this._action;
        }
        set action(value) {
            if (value instanceof Object) {
                if (typeof value.trigger === "string" &&
                    typeof value.callback === "function") {
                    if (this._triggerListener) {
                        this.container.removeEventListener(this.action.trigger, this._triggerListener);
                        this._triggered = false;
                        this._actionEvent = null;
                    }
                    this._action = value;

                    this._triggerListener = (evt) => {
                        this._actionEvent = evt;
                        this._triggered = true;
                        this._redetectMouseMove();
                        if (this.action?.callback) {
                          this._redetectMouseMove(); // call it again as the action may have altered the page
                        }
                    }
                    this.container.addEventListener(this.action.trigger, this._triggerListener);
                } else if (value.trigger !== undefined || value.callback !== undefined){ // allow empty action object
                    throw new Error("action must include two keys: trigger (String) and callback (function)!");
                }
            } else {
                throw new Error("action must be an object!");
            }
        }
        close() {
            if (this._triggerListener) {
              this.container.removeEventListener(this.action.trigger, this._triggerListener);
            }
            this.container.removeEventListener("mousemove", this._detectMouseMove);
            document.removeEventListener("scroll", this._onScroll);
            this.hoverBox.remove();
            this.hoverBoxInfo.remove();
            if (this.iframe) {
              this.iframe.remove();
            }
        }        
        _redetectMouseMove() {
            if (this._detectMouseMove && this._previousEvent) {
                this._detectMouseMove(this._previousEvent);
            }
        }
    }
    // export module
    if (typeof module !== 'undefined' && typeof module.exports !== 'undefined') {
        module.exports = ElementPicker;
    } else {
        window.ElementPicker = ElementPicker;
    }
})();
